File: bindings/go/libc_from_loaders.go
package memless

import "slices"

var (
	muslLoaders  = []string{"/lib/ld-musl-x86_64.so.1", "/lib/ld-musl-aarch64.so.1"}
	glibcLoaders = []string{"/lib64/ld-linux-x86-64.so.2", "/lib/ld-linux-aarch64.so.1"}
)

// libcFromLoaders names the libc by its dynamic loader; the musl loader wins,
// since gcompat installs a glibc one on Alpine.
func libcFromLoaders(exists func(string) bool) string {
	if slices.ContainsFunc(muslLoaders, exists) {
		return "musl"
	}
	if slices.ContainsFunc(glibcLoaders, exists) {
		return "glibc"
	}
	return ""
}

File: bindings/node/src/libc-from-loaders.js
'use strict';

const muslLoaders = ['/lib/ld-musl-x86_64.so.1', '/lib/ld-musl-aarch64.so.1'];
const glibcLoaders = ['/lib64/ld-linux-x86-64.so.2', '/lib/ld-linux-aarch64.so.1'];

// The dynamic loader tells the libc when ldd is missing, as in distroless
// images; the musl loader wins, since gcompat installs a glibc one on Alpine.
function libcFromLoaders(exists) {
  if (muslLoaders.some(exists)) {
    return 'musl';
  }
  return glibcLoaders.some(exists) ? 'glibc' : null;
}

module.exports = libcFromLoaders;
